### PR TITLE
Add downsampling metrics

### PR DESCRIPTION
**Pull request by mattrco**
When `compact` goes into its downsampling step, no app metrics are reported, which makes our dashboard look a bit bare for several hours (sometimes days if we have a backlog). Adding a couple of metrics means we can alert if no compactions or downsamples have happened in x period.

This is a little less clean than I'd like due to the two callers of `downsampleBucket` having slightly different contexts, however I think it's a reasonable compromise. Let me know what you think!

The metrics are labelled in the same way as compact metrics for ease of comparison.

## Changes

Add `thanos_compact_downsample_total` and `thanos_compact_downsample_failures_total` downsampling metrics.

## Verification

Tests are passing, will run with real workloads.

### CODE REVIEW

File: cmd/thanos/downsample.go [hunk 1]
```diff
@@ -31,20 +32,50 @@ import (
...
+	m.downsamples = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "thanos_compact_downsample_total",
+		Help: "Total number of downsampling attempts.",
+	}, []string{"group"})
+	m.downsampleFailures = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "thanos_compact_downsample_failures_total",
+		Help: "Total number of failed downsampling attempts.",
+	}, []string{"group"})
+
+	if err := reg.Register(m.downsamples); err != nil {
```
Comment (mattrco) [inline cmd/thanos/downsample.go]: This could use `MustRegister` (as done for other metrics) and skip the error check in the caller.

Comment (bwplotka) [reply cmd/thanos/downsample.go]: Yes, `MustRegister` is ok here

File: cmd/thanos/downsample.go [hunk 2]
```diff
@@ -31,20 +32,50 @@ import (
...
+		return nil, err
+	}
+	if err := reg.Register(m.downsampleFailures); err != nil {
```
Comment (bwplotka) [inline cmd/thanos/downsample.go]: Yup. `MustRegister` is fine here

**Comment by mattrco** [comment]
Thanks for reviewing, have updated to use `MustRegister` 👍

File: cmd/thanos/downsample.go [hunk 3]
```diff
@@ -31,20 +32,46 @@ import (
...
 
 	objStoreConfig := regCommonObjStoreFlags(cmd, "", true)
 
 	m[name] = func(g *run.Group, logger log.Logger, reg *prometheus.Registry, tracer opentracing.Tracer, _ bool) error {
-		return runDownsample(g, logger, reg, *dataDir, objStoreConfig)
+		return runDownsample(g, logger, reg, *httpAddr, *dataDir, objStoreConfig)
 	}
 }
 
+type DownsampleMetrics struct {
```
Comment (bwplotka) [inline cmd/thanos/downsample.go]: I think it can be private, right?

File: cmd/thanos/downsample.go [hunk 4]
```diff
@@ -65,6 +92,8 @@ func runDownsample(
 		}
 	}()
 
+	metrics := newDownsampleMetrics(reg)
```
Comment (GiedriusS) [inline cmd/thanos/downsample.go]: Could we use `extprom.WrapRegistererWithPrefix("thanos_compact_downsample", reg)` here?

File: cmd/thanos/downsample.go [hunk 5]
```diff
@@ -31,20 +32,46 @@ import (
...
 
+type DownsampleMetrics struct {
+	downsamples               *prometheus.CounterVec
+	downsampleFailures        *prometheus.CounterVec
+}
+
+func newDownsampleMetrics(reg *prometheus.Registry) *DownsampleMetrics {
+	m := new(DownsampleMetrics)
+
+	m.downsamples = prometheus.NewCounterVec(prometheus.CounterOpts{
```
Comment (GiedriusS) [inline cmd/thanos/downsample.go]: Then we would only have `total` here

File: cmd/thanos/downsample.go (hunk 4 above)
Comment (bwplotka) [reply cmd/thanos/downsample.go]: I think it makes sense but with `thanos_compact` prefix, but not a blocker.

**Comment by bwplotka** [comment]
Merging in this form, we can improve that later once we actually need to reuse this component.

**Comment by bwplotka** [comment]
Thanks @mattrco !

**Pull request merged**